src/mesh/boundary_info.C

```diff
@@ -1712,6 +1848,60 @@ BoundaryInfo::build_shellface_boundary_ids(std::vector<boundary_id_type> & b_ids
     }
 }
 
+void
+BoundaryInfo::transfer_boundary_ids_from_children(const Elem * const parent)
+{
+
+  // this is only needed when we allow boundary to be associated with children elements
+  // also, we only transfer the parent's boundary ids when we are actually coarsen the child element
+  if (!_children_on_boundary ||
+      !(!parent->active() && parent->refinement_flag() == Elem::COARSEN_INACTIVE))
+    return;
+
+  // We assume that edges can be divided ito two pieces, while triangles and
+  // quads can be divided into four smaller areas. This is double because we'll need
+  // to convert the ratio of the children with given boundary id to a double.
+  const double number_of_sides_on_children = std::pow(2, parent->dim()-1);
+
+  // In this case the input argument elem is the parent element. We need to check all of its sides
+  // to grab any potential boundary ids.
+  for (unsigned int side_i = 0; side_i < parent->n_sides(); ++side_i)
+  {
+    // An temporary storage to count how many times the children's boundaries occur. the general
+    // consensus is that if the boundary occurs more than once we propagate upon coarsening. Otherwise,
+    // it will get deleted.
+    std::map<unsigned short int, unsigned short int> boundary_counts;
+
+    for (const auto & child_i : make_range(parent->n_children()))
+    {
+      // We only need to check the children which share the side
+      if (parent->is_child_on_side(child_i, side_i))
+      {
+        // Fetching the boundary tags on the child's side
+        for (const auto & pr : as_range(_boundary_side_id.equal_range(parent->child_ptr(child_i))))
+        {
+          // Making sure we are on the same boundary
+          if (pr.second.first == side_i)
+            ++boundary_counts[pr.second.second];
+        }
+      }
+    }
+
+    // This is where the decision is made. If 50% of the children have the tags,
+    // we propagate them upwards upon coarsening. Otherwise, they are deleted.
+    for (const auto & boundary : boundary_counts)
+    {
+      if (boundary.second / number_of_sides_on_children > 0.5)
+        this->add_side(parent, side_i, boundary.first);
+      else
+        this->remove_side(parent, side_i, boundary.first);
+    }
+  }
+
+  for (const auto & child_i : make_range(parent->n_children()))
+    this->remove(parent->child_ptr(child_i));
```

Comment:
Isn't this redundant with the `remove()` calls in `MeshRefinement::_coarsen_elements`?

Reply:
I ended up keeping this and adding a condition for the removal in the coarsen loop. The reason behind this is that we might accidentally remove boundary ids before we transfer them back to the parent. Thus, when we allow boundaries on children we remove them after the transfer process.